pkg/pgmodel/migrate.go

```diff
@@ -105,16 +103,74 @@ func Migrate(db *pgxpool.Pool, versionInfo VersionInfo) (err error) {
 	if err != nil {
 		return fmt.Errorf("app version is not semver format, aborting migration")
 	}
-	if err := ensureVersionTable(db); err != nil {
+
+	mig := NewMigrator(db, migrations.MigrationFiles, tableOfContets)
+
+	err = mig.Migrate(appVersion)
+	if err != nil {
+		return fmt.Errorf("Error encountered during migration: %w", err)
+	}
+
+	_, extErr := conn.Exec(context.Background(), fmt.Sprintf(extensionInstall, extSchema))
+	if extErr != nil {
+		log.Warn("msg", "timescale_prometheus_extra extension not installed", "cause", extErr)
+		ExtensionIsInstalled = false
+	} else {
+		ExtensionIsInstalled = true
+	}
+
+	metadataUpdate(db, ExtensionIsInstalled, "version", versionInfo.Version)
+	metadataUpdate(db, ExtensionIsInstalled, "commit_hash", versionInfo.CommitHash)
+
+	return nil
+}
+
+type Migrator struct {
+	db       *pgxpool.Pool
+	sqlFiles http.FileSystem
+	toc      map[string][]string
+}
+
+func NewMigrator(db *pgxpool.Pool, sqlFiles http.FileSystem, toc map[string][]string) *Migrator {
+	return &Migrator{db: db, sqlFiles: sqlFiles, toc: toc}
+}
+
+func (t *Migrator) Migrate(appVersion semver.Version) error {
+	if err := ensureVersionTable(t.db); err != nil {
 		return fmt.Errorf("error ensuring version table: %w", err)
 	}
-	dbVersion, err := getDBVersion(db)
+
+	dbVersion, err := getDBVersion(t.db)
 	if err != nil {
 		return fmt.Errorf("failed to get the version from database: %w", err)
 	}
 
-	// If already at correct version, nothing to migrate.
+	// If already at correct version, nothing to migrate on proper release.
+	// On dev versions, idempotent files need to be reapplied.
```

Comment:
how come idempotent files need to be reapplied even if we're on the correct version?
(If it's just for testing, we should probably have a different flag to force the reapplication)

Reply:
This is just on dev/master versions. We don't want to require bumping versions for idempotent changes.

Reply:
Please add this clarification to the comment!